### PR TITLE
[feat] Introduce variable aliases

### PR DESCRIPTION
**Pull request by vkarak**
This practically extends on the field aliasing that we used to have and brings the same concept to variables. Users can now define a variable as an alias of another one. This is quite useful for transitional periods and deprecations when a variable is renamed. An aliased can be defined as follows:

```python
x = variable(int, value=3)
y = variable(alias=x)
```

The variable `y` may not have a default value, but it can be independently loggable. Any update of variable `x` will be visible to `y` and vice versa. Aliased variables can refer also to inherited variables.

This feature is useful for transitional periods when a variable is renamed and a deprecation is introduced and we want to keep both names for a while. The new variable can also be wrapped with `deprecate()` to mark a deprecation. This allows us to introduces variable aliases and deprecations effortlessly.

This commit also introduces implementation improvements on how variables are inherited. In particular, inherited variables are no more set redundantly as they are being inherited by classes lower in the hierarchy.

### CODE REVIEW

File: reframe/core/variables.py [hunk 1]
```diff
@@ -172,16 +172,40 @@ class MyRequiredTest(HelloTest):
...
+                 '_loggable', '_name', '_target')
+
+    __mutable_props = ('_default_value',)
 
     def __init__(self, *args, **kwargs):
         field_type = kwargs.pop('field', fields.TypedField)
-        self._default_value = kwargs.pop('value', Undefined)
+        alias = kwargs.pop('alias', None)
+
+        if alias and not isinstance(alias, TestVar):
```
Comment (victorusu) [inline reframe/core/variables.py:194]: Does it make sense to move these tests to a function?

**Review by victorusu — APPROVED** [review]
`lgtm`

File: reframe/core/variables.py (hunk 1 above)
Comment (vkarak) [reply reframe/core/variables.py]: Why? They are only used here.

**Comment by victorusu** [comment]
@teojgo and @ekouts, could you please take a look at this PR?

**Review by teojgo — APPROVED** [review]
lgtm, I only have a question. Is the `null` string the default one when the alias is not loggable? I couldn't find where it comes from

**Comment by vkarak** [comment]
> lgtm, I only have a question. Is the null string the default one when the alias is not loggable? I couldn't find where it comes from

The alias get their default value from their target. See the `_default_value` property.

**Comment by ekouts** [comment]
@jenkins-cscs retry dom

**Comment by teojgo** [comment]
> > lgtm, I only have a question. Is the null string the default one when the alias is not loggable? I couldn't find where it comes from
> 
> The alias get their default value from their target. See the `_default_value` property.

Ok, yeah it was difficult to find where the `null` comes from because in the end it comes from the `json.dumps`.

**Pull request merged**